### PR TITLE
Add permissions cache

### PR DESCRIPTION
**Pull request by mrbrowning**
This doesn't do much of interest right now, just planting a flag until the actual communication with Jack's vorenus-controller is fleshed out. For now, maintain token buckets in a cache per (directional) service pair that are reset on a one-second basis, removing a token for each attempted call. Going to tackle the gossiping between egress nodes next, which will probably inform a better choice for cache bound than Infinity.

### CODE REVIEW

File: node/hyperbahn/permissions_cache.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    self.channel = self.options.channel;
+    self.logger = self.options.logger;
+    self._intervalId = setInterval(
+        bucketResetCallback(self), BUCKET_RESET_DURATION
+    );
+
+    self.channel.statEvent.addListener(self.increment.bind(self));
+}
+
+inherits(PermissionsCache, LRUCache);
```
Comment (Raynos) [inline node/hyperbahn/permissions_cache.js]: Probably better to have `self.lru = new LRUCache()` then to inherit.

File: node/hyperbahn/permissions_cache.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+var NUM_TOKENS = 100;
+
+function PermissionsCache(options) {
+    if (!(this instanceof PermissionsCache)) {
+        return new PermissionsCache(options);
+    }
+    var self = this;
+
+    self.options = options;
+    PermissionsCache.super_.call(self, self.options);
```
Comment (Raynos) [inline node/hyperbahn/permissions_cache.js]: style nitpick: `LRUCache.call(self, options);`

File: node/hyperbahn/permissions_cache.js [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+// the actual rate limit from vorenus-controller.
+var NUM_TOKENS = 100;
+
+function PermissionsCache(options) {
+    if (!(this instanceof PermissionsCache)) {
+        return new PermissionsCache(options);
+    }
+    var self = this;
+
+    self.options = options;
```
Comment (Raynos) [inline node/hyperbahn/permissions_cache.js]: `self.options = options;` is bad pattern in tchannel; we should avoid it. just unpack things you need.

File: node/hyperbahn/permissions_cache.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    PermissionsCache.super_.call(self, self.options);
+
+    self.channel = self.options.channel;
+    self.logger = self.options.logger;
+    self._intervalId = setInterval(
+        bucketResetCallback(self), BUCKET_RESET_DURATION
```
Comment (Raynos) [inline node/hyperbahn/permissions_cache.js]: style:

```js
self._intervalId = setInterval(bucketReset, BUCKET_RESET_DURATION);

function bucketReset() {
    self.bucketReset();
}
```

File: node/hyperbahn/permissions_cache.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    );
+
+    self.channel.statEvent.addListener(self.increment.bind(self));
```
Comment (Raynos) [inline node/hyperbahn/permissions_cache.js]: bind() is banned.

```js
self.channel.statEvent.addListener(increment);

function increment(stat) {
    self.increment(stat);
}
```

File: node/hyperbahn/permissions_cache.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+inherits(PermissionsCache, LRUCache);
+
+PermissionsCache.prototype.clearBuckets = function clearBuckets() {
+    var self = this;
+    clearInterval(self._intervalId);
+    self.reset();
+};
+
+PermissionsCache.prototype.increment = function increment(stat) {
```
Comment (Raynos) [inline node/hyperbahn/permissions_cache.js]: nice trick hooking into the stats.

File: node/hyperbahn/permissions_cache.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+PermissionsCache.prototype.resetBucketTokens = function resetBucketTokens(key) {
+    // A pretty anemic method right now to be sure, but I suspect there might
+    // be more to this logic in the future.
+    var self = this;
+    self.set(key, NUM_TOKENS);
+};
+
+function bucketResetCallback(cache) {
+    return function() {
+        cache.keys().map(cache.resetBucketTokens.bind(cache));
```
Comment (Raynos) [inline node/hyperbahn/permissions_cache.js]: `bind()` is banned.

```js
var keys = cache.keys();
for (var i = 0; i < keys.length; i++) {
  cache.resetBucketTokens(keys[i]);
}
```

File: node/hyperbahn/service_proxy.js [hunk 8]
```diff
@@ -42,6 +43,10 @@ function ServiceDispatchHandler(options) {
 
     self.egressNodes.on('membershipChanged', onMembershipChanged);
 
+    self.permissionsCache = new PermissionsCache({
```
Comment (Raynos) [inline node/hyperbahn/service_proxy.js]: Let's not place this here. Let's just take `self.permissionsCache = options.permissionsCache;`

We can then instantiate it in autobahn/clients/index.js

**Comment by Raynos** [comment]
Mostly node/tchannel style and such; this looks reasonable on the surface. 

We do need to get rid of some of the slow paths though (.bind(), .map())

File: node/hyperbahn/permissions_cache.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+PermissionsCache.prototype.resetBucketTokens = function resetBucketTokens(key) {
+    // A pretty anemic method right now to be sure, but I suspect there might
+    // be more to this logic in the future.
+    var self = this;
+    self.set(key, NUM_TOKENS);
+};
+
+function bucketResetCallback(cache) {
+    return function() {
```
Comment (Raynos) [inline node/hyperbahn/permissions_cache.js]: style: all functions need names

File: node/hyperbahn/permissions_cache.js (hunk 3 above)
Comment (mrbrowning) [inline node/hyperbahn/permissions_cache.js]: Cool, I did pick it up from reading elsewhere in the code base, but that sounds preferable.

File: node/hyperbahn/permissions_cache.js (hunk 1 above)
Comment (mrbrowning) [inline node/hyperbahn/permissions_cache.js]: Then just treat `self.lru` as a public member and have referrers access that property directly to get call counts?

Comment (Raynos) [inline node/hyperbahn/permissions_cache.js]: that's upto you; all fields that do not start with `_` are public; it's up to you whether you want to add your own methods or not.

**Comment by Raynos** [comment]
lgtm.

File: node/test/permissions_cache.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        serviceName: 'server',
+        hasNoParent: true,
+        headers: {
+            cn: 'client',
+            as: 'raw'
+        }
+    }).send('echo', 'a', 'b', onResponse);
+    
+    function onResponse(err, res, arg2, arg3) {
+        assert.ifError(err);
```
Comment (kriskowal) [inline node/test/permissions_cache.js]: `if (err) return assert.end(err);`

**Comment by kriskowal** [comment]
This is good forward motion. Let’s get the flag planted.

**Pull request merged**